exec should wait for shim

dlespiau opened the pull request:
I don't really want to have to queue stdout/err data in the proxy.
    
    To solve that, let's introduce a new constraint. The data path:
    
      shim <-> proxy <-> hyperstart
    
    has to be fully setup before we allow newcontainer/execmd to execute a
    new process. While we advise to start the shim as early as possible and
    issue the newcontainer/execcmd command after that, we need to
    synchronize somewhere. The easier way of doing that is to stall the
    newcontainer/execcmd commands until we see the corresponding shim
    registering itself.
    
    Fixes: https://github.com/clearcontainers/proxy/issues/21
    Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

jodh-intel commented on vm.go:
```diff
@@ -380,6 +382,20 @@ func (vm *vm) SendMessage(hyper *api.Hyper) error {
+func (session *ioSession) WaitForShim() error {
+	select {
+	case <-session.shimConnected:
+	case <-time.After(waitForShimTimeout):
```
This seems like a reasonable timeout, but since it is a relatively long time, I think it could be useful to display a message every second or so that shows the status. Something like:

```
Waiting for shim to register itself with token %s (timeout in %d seconds)
```

That way, it'll be clear from the log what the proxy is doing and that it hasn't just "hung".

jodh-intel commented on proxy_test.go:
```diff
@@ -596,3 +604,57 @@ func TestShimSignal(t *testing.T) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		time.Sleep(20 * time.Millisecond)
```
Second occurence of `20 * time.Millisecond` - might be worth a variable for this?

dlespiau commented:
PR updated:

- Added log message indicating we're waiting for the shim (didn't implement the "every 1s" enough, that's probably over doing it)
- Added a const for the 20ms. Actually not too convinced about that one, it's really two different constants that happen to have the same value, but ¯\_(ツ)_/¯ (sorry I really wanted to use that emote!).

jodh-intel commented:
lgtm
